{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Group</th>\n",
       "      <th>SRR</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Study</th>\n",
       "      <th>PRJ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Macaca mulatta northern</td>\n",
       "      <td>mulatta</td>\n",
       "      <td>SRR4454026</td>\n",
       "      <td>SAMN05883679</td>\n",
       "      <td>SRS1762015</td>\n",
       "      <td>SRP092140</td>\n",
       "      <td>PRJNA345528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Macaca mulatta southern low altitude</td>\n",
       "      <td>mulatta</td>\n",
       "      <td>SRR4454020</td>\n",
       "      <td>SAMN05883709</td>\n",
       "      <td>SRS1762009</td>\n",
       "      <td>SRP092140</td>\n",
       "      <td>PRJNA345529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Macaca mulatta southern high altitude</td>\n",
       "      <td>mulatta</td>\n",
       "      <td>SRR4453966</td>\n",
       "      <td>SAMN05883736</td>\n",
       "      <td>SRS1761955</td>\n",
       "      <td>SRP092140</td>\n",
       "      <td>PRJNA345530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Macaca mulatta Indian</td>\n",
       "      <td>mulatta</td>\n",
       "      <td>SRR5628058</td>\n",
       "      <td>SAMN07168901</td>\n",
       "      <td>SRS2238957</td>\n",
       "      <td>SRP049547</td>\n",
       "      <td>PRJNA251548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Macaca fascicularis northern</td>\n",
       "      <td>fascicularis</td>\n",
       "      <td>fasno</td>\n",
       "      <td>SAMN00116341</td>\n",
       "      <td>SRS117874</td>\n",
       "      <td>SRP045755</td>\n",
       "      <td>PRJNA51411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Macaca fascicularis southern</td>\n",
       "      <td>fascicularis</td>\n",
       "      <td>fasso</td>\n",
       "      <td>SAMD00006158</td>\n",
       "      <td>DRS000787</td>\n",
       "      <td>DRP000438</td>\n",
       "      <td>PRJDB2038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Macaca fuscata</td>\n",
       "      <td>mulatta</td>\n",
       "      <td>DRR002233</td>\n",
       "      <td>SAMD00011919</td>\n",
       "      <td>DRS001583</td>\n",
       "      <td>DRP000620</td>\n",
       "      <td>PRJDB2459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Macaca fuscata</td>\n",
       "      <td>mulatta</td>\n",
       "      <td>fuscata2</td>\n",
       "      <td>SAMD00013516</td>\n",
       "      <td>DRS002017</td>\n",
       "      <td>DRP000657</td>\n",
       "      <td>PRJDB2648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Macaca thibethana</td>\n",
       "      <td>sinica</td>\n",
       "      <td>SRR1024051</td>\n",
       "      <td>SAMN02390221</td>\n",
       "      <td>SRS498543</td>\n",
       "      <td>SRP032525</td>\n",
       "      <td>PRJNA226187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Macaca assamensis</td>\n",
       "      <td>sinica</td>\n",
       "      <td>SRR2981114</td>\n",
       "      <td>SAMN04316321</td>\n",
       "      <td>SRS1196892</td>\n",
       "      <td>SRP067118</td>\n",
       "      <td>PRJNA305009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Macaca arctoides</td>\n",
       "      <td>fascicularis</td>\n",
       "      <td>SRR2981139</td>\n",
       "      <td>SAMN04316319</td>\n",
       "      <td>SRS1196879</td>\n",
       "      <td>SRP067118</td>\n",
       "      <td>PRJNA305009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Macaca arctoides</td>\n",
       "      <td>fascicularis</td>\n",
       "      <td>SRR2981140</td>\n",
       "      <td>SAMN04316320</td>\n",
       "      <td>SRS1196878</td>\n",
       "      <td>SRP067118</td>\n",
       "      <td>PRJNA305009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Macaca nemestrina</td>\n",
       "      <td>nemestrina</td>\n",
       "      <td>SRR5947292</td>\n",
       "      <td>SAMN07503410</td>\n",
       "      <td>SRS2436621</td>\n",
       "      <td>SRP115775</td>\n",
       "      <td>PRJNA398198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Macaca nemestrina</td>\n",
       "      <td>nemestrina</td>\n",
       "      <td>nemestrina2</td>\n",
       "      <td>SAMN10499717</td>\n",
       "      <td>SRS4092093</td>\n",
       "      <td>SRP171156</td>\n",
       "      <td>PRJNA507022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Macaca tonkeana</td>\n",
       "      <td>nemestrina</td>\n",
       "      <td>SRR5947293</td>\n",
       "      <td>SAMN07503429</td>\n",
       "      <td>SRS2441459</td>\n",
       "      <td>SRP115775</td>\n",
       "      <td>PRJNA398198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Macaca nigra</td>\n",
       "      <td>nemestrina</td>\n",
       "      <td>SRR5947294</td>\n",
       "      <td>SAMN07503430</td>\n",
       "      <td>SRS2436633</td>\n",
       "      <td>SRP115775</td>\n",
       "      <td>PRJNA398198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Macaca sylvanus</td>\n",
       "      <td>sylvanus</td>\n",
       "      <td>sylvanus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Macaca silenus</td>\n",
       "      <td>nemestrina</td>\n",
       "      <td>silenus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Papio anubis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR7588781</td>\n",
       "      <td>SAMN09710581</td>\n",
       "      <td>SRS3592457</td>\n",
       "      <td>SRP132725</td>\n",
       "      <td>PRJNA433868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Papio anubis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR8285768</td>\n",
       "      <td>SAMN10524549</td>\n",
       "      <td>SRS4110706</td>\n",
       "      <td>SRP132725</td>\n",
       "      <td>PRJNA433868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Papio anubis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN03073708</td>\n",
       "      <td>SRS742664</td>\n",
       "      <td>SRP026666</td>\n",
       "      <td>PRJNA54005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Species         Group          SRR  \\\n",
       "0                 Macaca mulatta northern       mulatta   SRR4454026   \n",
       "1    Macaca mulatta southern low altitude       mulatta   SRR4454020   \n",
       "2   Macaca mulatta southern high altitude       mulatta   SRR4453966   \n",
       "3                   Macaca mulatta Indian       mulatta   SRR5628058   \n",
       "4            Macaca fascicularis northern  fascicularis        fasno   \n",
       "5            Macaca fascicularis southern  fascicularis        fasso   \n",
       "6                          Macaca fuscata       mulatta    DRR002233   \n",
       "7                          Macaca fuscata       mulatta     fuscata2   \n",
       "8                       Macaca thibethana        sinica   SRR1024051   \n",
       "9                       Macaca assamensis        sinica   SRR2981114   \n",
       "10                       Macaca arctoides  fascicularis   SRR2981139   \n",
       "11                       Macaca arctoides  fascicularis   SRR2981140   \n",
       "12                      Macaca nemestrina    nemestrina   SRR5947292   \n",
       "13                      Macaca nemestrina    nemestrina  nemestrina2   \n",
       "14                        Macaca tonkeana    nemestrina   SRR5947293   \n",
       "15                           Macaca nigra    nemestrina   SRR5947294   \n",
       "16                        Macaca sylvanus      sylvanus     sylvanus   \n",
       "17                         Macaca silenus    nemestrina      silenus   \n",
       "18                           Papio anubis           NaN   SRR7588781   \n",
       "19                           Papio anubis           NaN   SRR8285768   \n",
       "20                           Papio anubis           NaN          NaN   \n",
       "\n",
       "       BioSample       Sample       Study          PRJ  \n",
       "0   SAMN05883679  SRS1762015   SRP092140   PRJNA345528  \n",
       "1   SAMN05883709   SRS1762009  SRP092140   PRJNA345529  \n",
       "2   SAMN05883736   SRS1761955   SRP092140  PRJNA345530  \n",
       "3   SAMN07168901   SRS2238957   SRP049547  PRJNA251548  \n",
       "4   SAMN00116341    SRS117874   SRP045755   PRJNA51411  \n",
       "5   SAMD00006158    DRS000787   DRP000438    PRJDB2038  \n",
       "6   SAMD00011919    DRS001583   DRP000620    PRJDB2459  \n",
       "7   SAMD00013516    DRS002017   DRP000657    PRJDB2648  \n",
       "8   SAMN02390221    SRS498543   SRP032525  PRJNA226187  \n",
       "9   SAMN04316321   SRS1196892   SRP067118  PRJNA305009  \n",
       "10  SAMN04316319   SRS1196879   SRP067118  PRJNA305009  \n",
       "11  SAMN04316320   SRS1196878   SRP067118  PRJNA305009  \n",
       "12  SAMN07503410   SRS2436621   SRP115775  PRJNA398198  \n",
       "13  SAMN10499717   SRS4092093   SRP171156  PRJNA507022  \n",
       "14  SAMN07503429   SRS2441459   SRP115775  PRJNA398198  \n",
       "15  SAMN07503430   SRS2436633   SRP115775  PRJNA398198  \n",
       "16           NaN          NaN         NaN          NaN  \n",
       "17           NaN          NaN         NaN          NaN  \n",
       "18  SAMN09710581   SRS3592457   SRP132725  PRJNA433868  \n",
       "19  SAMN10524549   SRS4110706   SRP132725  PRJNA433868  \n",
       "20  SAMN03073708    SRS742664   SRP026666   PRJNA54005  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"/moto/eaton/projects/macaques/metadata.csv\")\n",
    "df[[\"Species\", \"Group\", \"SRR\", \"BioSample\", \"Sample\", \"Study\", \"PRJ\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "for i in Chr20 Chr19 Chr18 Chr17 Chr16 Chr15 Chr14 Chr13 Chr12 Chr11 Chr10 Chr9 Chr8 Chr7 Chr6 Chr5 Chr4 Chr3 Chr2 Chr1; do\n",
    "    bcftools filter -e 'FORMAT/DP<10 | FORMAT/GQ<30 | FORMAT/GT=\"0/1\"' --set-GTs . --threads 12 /moto/eaton/projects/macaques/calls/$i.raw.vcf -O v -o /moto/eaton/projects/macaques/calls/$i.GQDP.vcf\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "for i in Chr20 Chr19 Chr18 Chr17 Chr16 Chr15 Chr14 Chr13 Chr12 Chr11 Chr10 Chr9 Chr8 Chr7 Chr6 Chr5 Chr4 Chr3 Chr2 Chr1; do\n",
    "    vcftools --recode --vcf /moto/eaton/projects/macaques/calls/$i.GQDP.vcf --minGQ 30 --minDP 10 --max-missing 0.4 --out /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "for i in Chr20 Chr19 Chr18 Chr17 Chr16 Chr15 Chr14 Chr13 Chr12 Chr11 Chr10 Chr9 Chr8 Chr7 Chr6 Chr5 Chr4 Chr3 Chr2 Chr1; do\n",
    "    python /moto/home/nsl2119/vcf2phylip/vcf2phylip.py \\\n",
    "        -i /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.recode.vcf --outgroup SRR8285768 \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate base\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr20 Chr19 Chr18 Chr17 Chr16 Chr15 Chr14 Chr13 Chr12 Chr11 Chr10 Chr9 Chr8 Chr7 Chr6 Chr5 Chr4 Chr3 Chr2 Chr1; do\n",
    "    python /moto/eaton/users/nsl2119/raxml_ascbias/ascbias.py -p /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.recode.min4.phy \\\n",
    "        -o /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr5 Chr4 Chr3 Chr2 Chr1; do\n",
    "    raxmlHPC-PTHREADS-SSE3 -s /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy -m ASC_GTRGAMMA \\\n",
    "        --asc-corr=lewis -f a -T 12 -N 100 -n $i -o SRR8285768 -w /moto/eaton/projects/macaques/calls/ -x 12345 -p 12345\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr10 Chr9 Chr8 Chr7 Chr6; do\n",
    "    raxmlHPC-PTHREADS-SSE3 -s /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy -m ASC_GTRGAMMA \\\n",
    "        --asc-corr=lewis -f a -T 12 -N 100 -n $i -o SRR8285768 -w /moto/eaton/projects/macaques/calls/ -x 12345 -p 12345\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr15 Chr14 Chr13 Chr12 Chr11; do\n",
    "    raxmlHPC-PTHREADS-SSE3 -s /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy -m ASC_GTRGAMMA \\\n",
    "        --asc-corr=lewis -f a -T 12 -N 100 -n $i -o SRR8285768 -w /moto/eaton/projects/macaques/calls/ -x 12345 -p 12345\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr20 Chr19 Chr18 Chr17 Chr16; do\n",
    "    raxmlHPC-PTHREADS-SSE3 -s /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy -m ASC_GTRGAMMA \\\n",
    "        --asc-corr=lewis -f a -T 12 -N 100 -n $i -o SRR8285768 -w /moto/eaton/projects/macaques/calls/ -x 12345 -p 12345\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To run RAxML on the whole genome variant sites, convert the PHYLIP to FASTA (`wget -P /moto/eaton/projects/macaques/scripts https://raw.githubusercontent.com/josephhughes/Sequence-manipulation/master/Phylip2Fasta.pl`)  and then concatenate with seqkit (`conda install -c bioconda seqkit`), then input into RAxML as usual (RAxML can take FASTA)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate base\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr20 Chr19 Chr18 Chr17 Chr16 Chr15 Chr14 Chr13 Chr12 Chr11 Chr10 Chr9 Chr8 Chr7 Chr6 Chr5 Chr4 Chr3 Chr2 Chr1; do\n",
    "    perl /moto/eaton/projects/macaques/scripts/Phylip2Fasta.pl /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy \\\n",
    "        /moto/eaton/projects/macaques/calls/$i.fasta\n",
    "    done\n",
    "seqkit concat --seq-type dna --threads 12 /moto/eaton/projects/macaques/calls/*.fasta \\\n",
    "    --out-file /moto/eaton/projects/macaques/calls/ALLCHRRAXML.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "raxmlHPC-PTHREADS-SSE3 -s /moto/eaton/projects/macaques/calls/ALLCHRRAXML.fasta -m ASC_GTRGAMMA \\\n",
    "        --asc-corr=lewis -f a -T 12 -N 100 -n ALLCHRRAXML -o SRR8285768 -w /moto/eaton/projects/macaques/calls/ -x 12345 -p 12345\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate vcfkit\n",
    "for i in Chr19 Chr18 Chr17; do\n",
    "    vk phylo fasta /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.recode.vcf > /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr19 Chr18 Chr17; do\n",
    "    perl /moto/eaton/projects/macaques/scripts/fasta2relaxedPhylip.pl -f /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.fasta -o /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.phy\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate base\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr19 Chr18 Chr17; do\n",
    "    python /moto/eaton/users/nsl2119/raxml_ascbias/ascbias.py -p /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.phy -o /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate py2\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "for i in Chr19 Chr18 Chr17; do\n",
    "    raxmlHPC-PTHREADS-SSE3 -s /moto/eaton/projects/macaques/calls/$i.GQDP.maxmiss.raxml.phy -m ASC_GTRGAMMA --asc-corr=lewis -f a -T 12 -N 100 -n $i -o SRR8285768 -w /moto/eaton/projects/macaques/calls/ -x 12345 -p 12345\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create a PHYLIP file of all the variant sites for the whole genome, we concatenate with seqkit (`conda install -c bioconda seqkit`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m read file: /moto/eaton/projects/macaques/calls/Chr10.GQDP.maxmiss.raxml.phy\n",
      "\u001b[31m[ERRO]\u001b[0m fastx: invalid FASTA/Q format\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate base\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "seqkit concat --seq-type dna --threads 12 /moto/eaton/projects/macaques/calls/*.GQDP.maxmiss.raxml.phy --out-file /moto/eaton/projects/macaques/calls/ALLCHRRAXML.phy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "source ~/.bashrc\n",
    "conda activate test\n",
    "export LC_ALL=\"en_US.UTF-8\"\n",
    "iqtree -s /moto/eaton/projects/macaques/calls/Chr20.GQDP.maxmiss.raxml.phy -o SRR8285768 -st DNA -m MFP+ASC -bb 5000 -alrt 5000 -nt 12 -mem 96G -redo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "iqtree -s /moto/eaton/projects/macaques/calls/vcftoolsfilterinvardel.phy -m MFP+ASC -bb 5000 -alrt 5000 -nt 12 -mem 96gb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree -s /moto/eaton/projects/macaques/mapped/Chr19/calls/vcftoolsfilter.phy -m MFP+ASC -nt 12 -n 10000 -mem 64gb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##on desktop\n",
    "scp nsl2119@moto.rcs.columbia.edu:/moto/eaton/projects/macaques/calls/RAxML_bestTree.vcftoolsfilter /home/nicolas/Desktop/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python py2",
   "language": "python",
   "name": "py2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
